apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -530,297 +533,605 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
                     {
                         _event.Set();
                     }
-                })));
+                }));
+                _registerTaskFn(task);
+            }
+        }
+
+        // internal for testing.
+        internal class RequeueableQueue<T>
+        {
+            private readonly Queue<T> _q = new Queue<T>();
+            private readonly LinkedList<Queue<T>> _qs = new LinkedList<Queue<T>>();
+            private int _requeueCount = 0;
+
+            public void Enqueue(T item) => _q.Enqueue(item);
+            public void Enqueue(IEnumerable<T> items)
+            {
+                foreach (var item in items)
+                {
+                    _q.Enqueue(item);
+                }
+            }
+            public void Requeue(IEnumerable<T> items)
+            {
+                var q = new Queue<T>(items);
+                _qs.AddLast(q);
+                _requeueCount += q.Count;
+            }
+
+            public int Count => _q.Count + _requeueCount;
+
+            public List<T> Dequeue(int maxCount, Predicate<T> includeFn)
```

Comment:
It feels like this could probably be refactored into a private "DequeueOne" method and then a while loop that calls it. I'm okay with it as-is for the moment, but I think it could be simpler. Maybe try that later on?